File: windows/GlassTranslate.Windows/Services/TranslationClient.cs
using System.Net;
using System.Globalization;
using System.Text;
using System.Text.Json;
using GlassTranslate.Windows.Models;

namespace GlassTranslate.Windows.Services;

public sealed class TranslationClient
{
    private static readonly HttpClient Http = new() { Timeout = TimeSpan.FromSeconds(30) };

    public async Task<string> TranslateAsync(string source, SettingsStore settings, CancellationToken cancellationToken)
    {
        return await TranslateWithMyMemoryAsync(source, settings, cancellationToken);
    }

    private static async Task<string> TranslateWithMyMemoryAsync(
        string source,
        SettingsStore settings,
        CancellationToken cancellationToken)
    {
        var sourceLanguage = DetectLanguageCode(source);
        var targetLanguage = TargetLanguageCode(settings.TargetLanguage, sourceLanguage);
        if (sourceLanguage == targetLanguage) return source;

        var translations = new List<string>();
        foreach (var chunk in Utf8Chunks(source, 450))
        {
            var url = new StringBuilder("https://api.mymemory.translated.net/get?q=")
                .Append(Uri.EscapeDataString(chunk))
                .Append("&langpair=")
                .Append(Uri.EscapeDataString($"{sourceLanguage}|{targetLanguage}"))
                .Append("&mt=1");
            if (!string.IsNullOrWhiteSpace(settings.FreeServiceEmail))
                url.Append("&de=").Append(Uri.EscapeDataString(settings.FreeServiceEmail.Trim()));

            using var request = new HttpRequestMessage(HttpMethod.Get, url.ToString());
            request.Headers.UserAgent.ParseAdd("Wonder/0.3");
            using var response = await Http.SendAsync(request, cancellationToken);
            var body = await response.Content.ReadAsStringAsync(cancellationToken);
            if (!response.IsSuccessStatusCode)
                throw new InvalidOperationException($"免费翻译服务请求失败（{(int)response.StatusCode}）。");
            try
            {
                using var document = JsonDocument.Parse(body);
                var root = document.RootElement;
                if (root.TryGetProperty("quotaFinished", out var quota) && quota.ValueKind == JsonValueKind.True)
                    throw new InvalidOperationException("今日免费额度已用完，可在设置中填写邮箱提升额度。");
                if (root.TryGetProperty("responseStatus", out var status) && status.TryGetInt32(out var code) && code >= 400)
                {
                    var details = root.TryGetProperty("responseDetails", out var detail) ? detail.GetString() : "请求被拒绝";
                    throw new InvalidOperationException($"免费翻译服务：{details}");
                }
                var translated = root.GetProperty("responseData").GetProperty("translatedText").GetString()?.Trim();
                if (string.IsNullOrWhiteSpace(translated)) throw new InvalidOperationException("免费翻译服务返回了空结果。");
                translations.Add(WebUtility.HtmlDecode(translated));
            }
            catch (JsonException)
            {
                throw new InvalidOperationException("免费翻译服务返回了无法识别的数据。");
            }
        }
        return string.Join(source.Contains('\n') ? "\n" : " ", translations);
    }

    internal static string DetectLanguageCode(string text)
    {
        foreach (var character in text)
        {
            if (character is >= '\u3040' and <= '\u30ff') return "ja";
            if (character is >= '\uac00' and <= '\ud7af') return "ko";
            if (character is >= '\u0400' and <= '\u04ff') return "ru";
        }
        return text.Any(character => character is >= '\u3400' and <= '\u9fff') ? "zh-CN" : "en";
    }

    internal static string TargetLanguageCode(TargetLanguage target, string source) => target switch
    {
        TargetLanguage.Automatic => source.StartsWith("zh", StringComparison.OrdinalIgnoreCase) ? "en" : "zh-CN",
        TargetLanguage.SimplifiedChinese => "zh-CN",
        TargetLanguage.English => "en",
        TargetLanguage.Japanese => "ja",
        TargetLanguage.Korean => "ko",
        TargetLanguage.French => "fr",
        TargetLanguage.German => "de",
        TargetLanguage.Spanish => "es",
        _ => "zh-CN"
    };

    internal static IReadOnlyList<string> Utf8Chunks(string text, int maximumBytes)
    {
        if (Encoding.UTF8.GetByteCount(text) <= maximumBytes) return [text];
        var chunks = new List<string>();
        var current = new StringBuilder();
        var currentBytes = 0;
        var elements = StringInfo.GetTextElementEnumerator(text);
        while (elements.MoveNext())
        {
            var value = elements.GetTextElement();
            var bytes = Encoding.UTF8.GetByteCount(value);
            if (currentBytes + bytes > maximumBytes && current.Length > 0)
            {
                chunks.Add(current.ToString());
                current.Clear();
                currentBytes = 0;
            }
            current.Append(value);
            currentBytes += bytes;
        }
        if (current.Length > 0) chunks.Add(current.ToString());
        return chunks;
    }

}

File: macos/Sources/GlassTranslate/TranslationClient.swift
import Foundation
#if canImport(FoundationNetworking)
import FoundationNetworking
#endif

struct TranslationClient: Sendable {
    struct Configuration: Sendable {
        let sourceLanguage: SourceLanguage
        let targetLanguage: TargetLanguage
        let freeServiceEmail: String
    }

    private struct MyMemoryResponse: Decodable, Sendable {
        struct ResponseData: Decodable, Sendable {
            let translatedText: String
        }
        let responseData: ResponseData
        let responseStatus: Int?
        let responseDetails: String?
        let quotaFinished: Bool?
    }

    let session: URLSession

    init(session: URLSession = .shared) {
        self.session = session
    }

    func translate(_ source: String, configuration: Configuration) async throws -> String {
        try await translateWithMyMemory(source, configuration: configuration)
    }

    private func translateWithMyMemory(_ source: String, configuration: Configuration) async throws -> String {
        let sourceLanguage = configuration.sourceLanguage.languageCode ?? Self.detectedLanguageCode(source)
        let targetLanguage = Self.targetLanguageCode(configuration.targetLanguage, source: sourceLanguage)
        if sourceLanguage == targetLanguage { return source }

        var translations: [String] = []
        for chunk in Self.utf8Chunks(source, maximumBytes: 450) {
            var components = URLComponents(string: "https://api.mymemory.translated.net/get")!
            var queryItems = [
                URLQueryItem(name: "q", value: chunk),
                URLQueryItem(name: "langpair", value: "\(sourceLanguage)|\(targetLanguage)"),
                URLQueryItem(name: "mt", value: "1")
            ]
            let email = configuration.freeServiceEmail.trimmingCharacters(in: .whitespacesAndNewlines)
            if !email.isEmpty { queryItems.append(URLQueryItem(name: "de", value: email)) }
            components.queryItems = queryItems
            guard let url = components.url else { throw TranslationError.invalidEndpoint }

            var request = URLRequest(url: url)
            request.timeoutInterval = 20
            request.setValue("Wonder/0.3", forHTTPHeaderField: "User-Agent")
            let (data, response) = try await session.data(for: request)
            guard let http = response as? HTTPURLResponse else { throw TranslationError.invalidResponse }
            guard 200..<300 ~= http.statusCode else {
                throw TranslationError.freeService("请求失败（\(http.statusCode)）")
            }
            guard let result = try? JSONDecoder().decode(MyMemoryResponse.self, from: data) else {
                throw TranslationError.invalidResponse
            }
            if result.quotaFinished == true {
                throw TranslationError.freeService("今日免费额度已用完，可在设置中填写邮箱提升额度。")
            }
            if let status = result.responseStatus, status >= 400 {
                throw TranslationError.freeService(result.responseDetails ?? "请求被拒绝")
            }
            let translated = result.responseData.translatedText.trimmingCharacters(in: .whitespacesAndNewlines)
            guard !translated.isEmpty else { throw TranslationError.emptyResult }
            translations.append(translated)
        }
        return translations.joined(separator: source.contains("\n") ? "\n" : " ")
    }

    static func detectedLanguageCode(_ text: String) -> String {
        for scalar in text.unicodeScalars {
            switch scalar.value {
            case 0x3040...0x30FF: return "ja"
            case 0xAC00...0xD7AF: return "ko"
            case 0x0400...0x04FF: return "ru"
            default: continue
            }
        }
        if text.unicodeScalars.contains(where: { 0x3400...0x9FFF ~= $0.value }) { return "zh-CN" }
        return "en"
    }

    static func targetLanguageCode(_ target: TargetLanguage, source: String) -> String {
        switch target {
        case .automatic: source.hasPrefix("zh") ? "en" : "zh-CN"
        case .simplifiedChinese: "zh-CN"
        case .traditionalChinese: "zh-TW"
        case .english: "en-US"
        case .englishUK: "en-GB"
        case .japanese: "ja"
        case .korean: "ko"
        case .french: "fr"
        case .german: "de"
        case .spanish: "es"
        case .russian: "ru"
        case .italian: "it"
        case .portuguese: "pt"
        }
    }

    static func utf8Chunks(_ text: String, maximumBytes: Int) -> [String] {
        guard text.utf8.count > maximumBytes else { return [text] }
        var chunks: [String] = []
        var current = ""
        var currentBytes = 0
        for character in text {
            let value = String(character)
            let bytes = value.utf8.count
            if currentBytes + bytes > maximumBytes, !current.isEmpty {
                chunks.append(current)
                current = ""
                currentBytes = 0
            }
            current.append(character)
            currentBytes += bytes
        }
        if !current.isEmpty { chunks.append(current) }
        return chunks
    }
}
